### ISSUE TITLE
Long circuit generation time

### ISSUE THREAD

**Issue by sanchopansa** [akosba/xjsnark#24]
Hi Ahmed,

I've been working on an implementation of the Pedersen hash function (over the baby_jubjub curve) in xjsnark and for some reason the circuit generation time for this is _much_ longer than I expected.

For example, for a Pedersen circuit with around 42K constraints, I've clocked the generation time at around 27 seconds on my laptop. For comparison, I have another circuit (doing ECDSA signature verification over secp256k1), which consists of 2.2M constraints and there the circuit generation time is only 9 seconds, so the overall difference is ~160x.

I am attaching the mps file here for reference. Do you have any idea why this is happening?
[pedersenTest.tar.gz](https://github.com/akosba/xjsnark/files/3278103/pedersenTest.tar.gz)

**Comment by sanchopansa** [comment]
For anyone reading this, please note that the above implementation of the Pedersen hash in not secure, as it's using the same generator point for the lookup table instead of using a new generator point every 189 bits.

**Comment by akosba** [comment]
Hi Valentin,

The number of constraints is not the only factor that affects the circuit generation time, and the comparison you mentioned does not include all the factors. Although the number of constraints is the primary factor we care about for the cryptographic keygen and proof computation, during the circuit generation stage itself, other metrics will come into play. For example, your code seems to have a lot of operations with constant BigInteger operands that are resolved within the compilation time, e.g., createLookupTable(). This does not seem to depend on the variable statement, and only deals with predetermined values. Intuitively this should not take a lot of time, but this will be slower than normal here. This is because the code has to go through many of the same methods/checks that are called when the inputs are variable. Also, addition and other cheap operations can play a role for the same reason. 
In the case of secp256k1 circuit, many of the constraints are just bit constraints, and hundreds of bit constraints are just generated in one method call. In the case of your code, it's different. I think your code might have many high-level operator calls in comparison.

So in short, the circuit generation time will be affected by the number of operations you call, even if they do not result in constraints. That said, I am sure there is room for improvement, especially in the way the back end deals with BigInteger operations/objects. I will try to give a more detailed look later to reduce the time of circuit generation in scenarios similar to yours.

**Comment by akosba** [comment]
Hi Valentin,

The number of constraints is not the only factor that affects the circuit generation time. Although the number of constraints is the primary factor we care about for the cryptographic keygen and proof computation, during the circuit generation stage itself, other metrics will come into play. For example, your code seems to have a lot of operations with constant BigInteger operands that are resolved within the compilation time, e.g., createLookupTable(). This does not seem to depend on the variable statement, and only deals with predetermined values. Intuitively this should not take a lot of time, but this will be slower than normal here. This is because the code has to go through many of the same methods/checks that are called when the inputs are variable. Also, addition and other cheap operations can play a role for the same reason.
In the case of secp256k1 circuit, many of the constraints are just bit constraints, and hundreds of bit constraints are just generated in one method call. In the case of your code, it's different. I think your code might have many high-level operator calls in comparison.

So in short, the circuit generation time will be affected by the number of operations you call, even if they do not result in constraints. That said, I am sure there is room for improvement, especially in the way the back end deals with BigInteger operations/objects. I will try to give a more detailed look later to reduce the time of circuit generation in scenarios similar to yours.

**Comment by akosba** [comment]
A couple of notes:
- I think it's possible to reduce the running time of your circuit generation by replacing 

```
for (int row = 0; row < 63; row++) { 
  F_p[] gen = BabyJubJub.scalarMultEdwards(multiplier, PedersenTest.generator_x, PedersenTest.generator_y); 
  table[row][0] = gen[0]; 
  table[row][1] = gen[1]; 
  multiplier = F_p(16) * multiplier; 
}
```
with a call to a native java method that returns a 2D array of precomputed BigInteger constants, which then you could assign to the table array directly. This has the downside of writing the same logic twice.

- The second problem that I highlighted in #25 might be playing a role in the long generation time as well, as it could have lead to more operations with constant operands.
